scipy/stats/_bootstrap.py

```diff
@@ -5,43 +5,103 @@
-    n = sample.shape[-1]
-    batch_nominal = batch or n
+class _PairedStatistic:
+    """
+    Object to wrap a paired sample statistic for use with `bootstrap`.
+
+    To bootstrap with paired-sample statistics, we resample the indices of
+    the observations. This class wraps a user-provided statistic and user-
+    provided data into a callable that accepts indices, rather than data,
+    as the first argument.
+    """
+    def __init__(self, data, statistic):
+        self.data = data
+        self.statistic = statistic
+
+    def __call__(self, i, axis=-1):
+        data = [sample[..., i] for sample in self.data]
+        return self.statistic(*data, axis=axis)
+
+
+class _StatisticLastAxis:
+    """
+    Object to wrap a statistic and call it along the last axis.
+    """
+    def __init__(self, statistic, default_axis=-1):
+        self.statistic = statistic
+        self.default_axis = default_axis
+
+    def __call__(self, data):
+        return self.statistic(*data, axis=self.default_axis)
+
+
+class _BootstrapDistribution:
+    def __init__(self, batch_nominal, n_resamples, data, statistic, rng):
+        self.batch_nominal = batch_nominal
+        self.n_resamples = n_resamples
+        self.data = data
+        self.statistic = statistic
+        self.rng = rng
 
-    for k in range(0, n, batch_nominal):
+    def __call__(self, k):
+        batch_actual = min(self.batch_nominal, self.n_resamples-k)
+        # Generate resamples
+        resampled_data = []
+        for sample in self.data:
+            resample = _bootstrap_resample(sample,
+                                           n_resamples=batch_actual,
+                                           random_state=self.rng)
+            resampled_data.append(resample)
+        return self.statistic(resampled_data)
+
+
+class _JackknifeDistribution:
+    def __init__(self, batch_nominal, n, sample, statistic):
+        self.batch_nominal = batch_nominal
+        self.n = n
+        self.sample = sample
+        self.statistic = statistic
+
+    def __call__(self, k):
         # col_start:col_end are the observations to remove
-        batch_actual = min(batch_nominal, n-k)
+        batch_actual = min(self.batch_nominal, self.n-k)
 
         # jackknife - each row leaves out one observation
-        j = np.ones((batch_actual, n), dtype=bool)
+        j = np.ones((batch_actual, self.n), dtype=bool)
         np.fill_diagonal(j[:, k:k+batch_actual], False)
-        i = np.arange(n)
-        i = np.broadcast_to(i, (batch_actual, n))
-        i = i[j].reshape((batch_actual, n-1))
-
-        resamples = sample[..., i]
-        yield resamples
+        i = np.arange(self.n)
+        i = np.broadcast_to(i, (batch_actual, self.n))
+        i = i[j].reshape((batch_actual, self.n-1))
+        resamples = self.sample[..., i]
+        return self.statistic((resamples,))
 
 
 def _bootstrap_resample(sample, n_resamples=None, random_state=None):
```

Comment:
For large problems, a substantial amount of time is spent in this function. I assume that sharing a `random_state` between the workers could produce a bottleneck, but that doesn't seem to account for much of the time when `workers=1`; the resampling itself `resamples = sample [..., i]` takes more time. Is there a bottleneck in sharing the memory?